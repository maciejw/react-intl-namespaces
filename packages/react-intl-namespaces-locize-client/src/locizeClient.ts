import { IntlNamespaces } from 'react-intl-namespaces';

import {
  NamespaceResource,
  ResourceServer,
} from 'react-intl-namespaces/contracts';

export class LocizeClient implements ResourceServer {
  private options: LocizeClient.ApiUrls & LocizeClient.RequiredOptions;
  private globalFetch: GlobalFetch;
  constructor(
    globalFetch: GlobalFetch,
    options: Partial<LocizeClient.ApiUrls> & LocizeClient.RequiredOptions,
  ) {
    this.globalFetch = this.configureFetch(globalFetch);
    this.options = {
      ...LocizeClient.getDefaults(),
      ...options,
    };
  }

  public async getLanguages(): Promise<string[]> {
    const response = await this.globalFetch.fetch(
      this.options.languagesUrl(this.options),
      {
        method: 'GET',
      },
    );
    return await response.json();
  }
  public async getNamespace(ns: string): Promise<NamespaceResource> {
    const params = { ...this.options, ns };
    const response = await this.globalFetch.fetch(
      this.options.namespaceUrl(params),
      {
        method: 'GET',
      },
    );
    const result = await response.json();

    const resource = IntlNamespaces.MessageConverter.flattenTree(result);

    return resource;
  }
  public async addMissing(
    ns: string,
    missingResources: NamespaceResource,
  ): Promise<void> {
    const params = { ...this.options, ns };
    const response = await this.globalFetch.fetch(
      this.options.missingUrl(params),
      {
        body: JSON.stringify(missingResources),
        method: 'POST',
      },
    );
    return;
  }
  public async updateModified(
    ns: string,
    modifiedResources: NamespaceResource,
    replace: boolean = false,
  ): Promise<void> {
    const params = { ...this.options, ns };
    let url = this.options.updateUrl(params);
    if (replace) {
      url += '?replace=true';
    }
    const response = await this.globalFetch.fetch(url, {
      body: JSON.stringify(modifiedResources),
      method: 'POST',
    });
    return;
  }

  private configureFetch(globalFetch: GlobalFetch): GlobalFetch {
    const fetch = async (input: RequestInfo, init: RequestInit = {}) => {
      const { headers, ...rest } = init;

      const defaults: RequestInit = {};
      defaults.headers = new Headers(headers);
      defaults.headers.append('Authorization', `Bearer ${this.options.apiKey}`);
      defaults.headers.append('Accept', `application/json`);
      defaults.headers.append('Content-Type', `application/json`);
      const response = await globalFetch.fetch(input, { ...defaults, ...rest });

      return response;
    };
    return { fetch };
  }
}

export namespace LocizeClient {
  export function getDefaults(): LocizeClient.ApiUrls {
    return {
      languagesUrl: ({ projectId }) =>
        `https://api.locize.io/languages/${projectId}`,
      missingUrl: ({ projectId, version, referenceLanguage, ns }) =>
        `https://api.locize.io/missing/${projectId}/${version}/${referenceLanguage}/${ns}`,
      namespaceUrl: ({ projectId, version, referenceLanguage, ns }) =>
        `https://api.locize.io/${projectId}/${version}/${referenceLanguage}/${ns}`,
      updateUrl: ({ projectId, version, referenceLanguage, ns }) =>
        `https://api.locize.io/update/${projectId}/${version}/${referenceLanguage}/${ns}`,
      version: 'latest',
    };
  }
  export interface RequiredOptions {
    apiKey: string;
    referenceLanguage: string;
    projectId: string;
  }
  export interface ApiUrls {
    languagesUrl: (params: { projectId: string }) => string;
    missingUrl: (
      params: {
        projectId: string;
        version: string;
        referenceLanguage: string;
        ns: string;
      },
    ) => string;
    namespaceUrl: (
      params: {
        projectId: string;
        version: string;
        referenceLanguage: string;
        ns: string;
      },
    ) => string;
    updateUrl: (
      params: {
        projectId: string;
        version: string;
        referenceLanguage: string;
        ns: string;
      },
    ) => string;
    version: string;
  }
}
