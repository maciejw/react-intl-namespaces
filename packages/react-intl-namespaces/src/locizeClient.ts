import { NamespaceResource, ResourceServer } from './namespaceTypes';

export class LocizeClient implements ResourceServer {
  private options: LocizeClient.ApiUrls & LocizeClient.RequiredOptions;
  private globalFetch: GlobalFetch;
  constructor(
    globalFetch: GlobalFetch,
    options: Partial<LocizeClient.ApiUrls> & LocizeClient.RequiredOptions,
  ) {
    this.globalFetch = this.configureFetch(globalFetch);
    this.options = {
      ...options,
      ...LocizeClient.getDefaults(),
    };
  }

  public async getLanguages(): Promise<string[]> {
    const response = await this.globalFetch.fetch(
      this.options.languagesUrl(this.options),
      {
        method: 'GET',
      },
    );
    return await response.json();
  }
  public async getNamespace(ns: string): Promise<NamespaceResource> {
    const params = { ...this.options, ns };
    const response = await this.globalFetch.fetch(
      this.options.namespaceUrl(params),
      {
        method: 'GET',
      },
    );
    return await response.json();
  }
  public async addMissing(
    ns: string,
    missingResources: NamespaceResource,
  ): Promise<void> {
    const params = { ...this.options, ns };
    const response = await this.globalFetch.fetch(
      this.options.missingUrl(params),
      {
        body: JSON.stringify(missingResources),
        method: 'POST',
      },
    );
    return;
  }
  public async updateModified(
    ns: string,
    modifiedResources: NamespaceResource,
    replace: boolean = false,
  ): Promise<void> {
    const params = { ...this.options, ns };
    let url = this.options.updateUrl(params);
    if (replace) {
      url += '?replace=true';
    }
    const response = await this.globalFetch.fetch(url, {
      body: JSON.stringify(modifiedResources),
      method: 'POST',
    });
    return;
  }

  private configureFetch(globalFetch: GlobalFetch): GlobalFetch {
    const fetch = async (input: RequestInfo, init: RequestInit = {}) => {
      const { headers, ...rest } = init;

      const defaults: RequestInit = {};
      defaults.headers = new Headers(headers);
      defaults.headers.append('Authorization', `Bearer ${this.options.apiKey}`);
      defaults.headers.append('Accept', `application/json`);
      defaults.headers.append('Content-Type', `application/json`);
      const response = await globalFetch.fetch(input, { ...defaults, ...rest });

      return response;
    };
    return { fetch };
  }
}

export namespace LocizeClient {
  export function getDefaults(): LocizeClient.ApiUrls {
    return {
      languagesUrl: ({ projectId }) =>
        `https://api.locize.io/languages/${projectId}`,
      missingUrl: ({ projectId, version, lng, ns }) =>
        `https://api.locize.io/missing/${projectId}/${version}/${lng}/${ns}`,
      namespaceUrl: ({ projectId, version, lng, ns }) =>
        `https://api.locize.io/${projectId}/${version}/${lng}/${ns}`,
      updateUrl: ({ projectId, version, lng, ns }) =>
        `https://api.locize.io/update/${projectId}/${version}/${lng}/${ns}`,
      version: 'latest',
    };
  }
  export interface RequiredOptions {
    apiKey: string;
    lng: string;
    projectId: string;
  }
  export interface ApiUrls {
    languagesUrl: (params: { projectId: string }) => string;
    missingUrl: (
      params: {
        projectId: string;
        version: string;
        lng: string;
        ns: string;
      },
    ) => string;
    namespaceUrl: (
      params: {
        projectId: string;
        version: string;
        lng: string;
        ns: string;
      },
    ) => string;
    updateUrl: (
      params: {
        projectId: string;
        version: string;
        lng: string;
        ns: string;
      },
    ) => string;
    version: string;
  }
}
