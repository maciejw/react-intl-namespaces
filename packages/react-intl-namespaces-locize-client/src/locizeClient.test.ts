import { LocizeClient } from './locizeClient';

describe('LocizeClient', () => {
  it('should get languages', async () => {
    const fetchMock = jest.fn();
    fetchMock.mockReturnValue(
      new Response(JSON.stringify({ pl: {}, en: {} }), { status: 200 }),
    );
    const backend = new LocizeClient(
      { fetch: fetchMock },
      { apiKey: 'apikey', projectId: 'projectid', referenceLanguage: 'en' },
    );
    const result = await backend.getLanguages();

    expect(result).toMatchSnapshot();
    expect(fetchMock).toMatchSnapshot();
  });

  it('should get namespace', async () => {
    const fetchMock = jest.fn();

    fetchMock.mockReturnValue(
      new Response(
        JSON.stringify({
          prefix: {
            res1: 'val2',
          },
          res1: 'val1',
        }),
      ),
    );
    const backend = new LocizeClient(
      { fetch: fetchMock },
      { apiKey: 'apikey', projectId: 'projectid', referenceLanguage: 'en' },
    );

    const result = await backend.getNamespace('ns1');

    expect(result).toMatchSnapshot();
    expect(fetchMock).toMatchSnapshot();
  });
  it('should add missing', async () => {
    const fetchMock = jest.fn();
    fetchMock.mockReturnValue(new Response(undefined, { status: 200 }));
    const backend = new LocizeClient(
      { fetch: fetchMock },
      { apiKey: 'apikey', projectId: 'projectid', referenceLanguage: 'en' },
    );

    await backend.addMissing('ns1', { res1: 'val' });

    expect(fetchMock).toMatchSnapshot();
  });
  it('should update modified', async () => {
    const fetchMock = jest.fn();
    fetchMock.mockReturnValue(new Response(undefined, { status: 200 }));
    const backend = new LocizeClient(
      { fetch: fetchMock },
      { apiKey: 'apikey', projectId: 'projectid', referenceLanguage: 'en' },
    );

    await backend.updateModified('ns1', { res1: 'val' });

    expect(fetchMock).toMatchSnapshot();
  });
  it('should replace', async () => {
    const fetchMock = jest.fn();
    fetchMock.mockReturnValue(new Response(undefined, { status: 200 }));
    const backend = new LocizeClient(
      { fetch: fetchMock },
      { apiKey: 'apikey', projectId: 'projectid', referenceLanguage: 'en' },
    );

    await backend.updateModified('ns1', { res1: 'val' }, true);

    expect(fetchMock).toMatchSnapshot();
  });
});
