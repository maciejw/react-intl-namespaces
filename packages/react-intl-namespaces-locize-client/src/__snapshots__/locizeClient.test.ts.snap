// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocizeClient should add missing 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.locize.io/missing/projectid/latest/en/ns1",
      Object {
        "body": "{\\"res1\\":\\"val\\"}",
        "headers": Headers {
          "map": Object {
            "accept": "application/json",
            "authorization": "Bearer apikey",
            "content-type": "application/json",
          },
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`LocizeClient should get languages 1`] = `
Array [
  "pl",
  "en",
]
`;

exports[`LocizeClient should get languages 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.locize.io/languages/projectid",
      Object {
        "headers": Headers {
          "map": Object {
            "accept": "application/json",
            "authorization": "Bearer apikey",
            "content-type": "application/json",
          },
        },
        "method": "GET",
      },
    ],
  ],
}
`;

exports[`LocizeClient should get namespace 1`] = `
Object {
  "prefix.res1": "val2",
  "res1": "val1",
}
`;

exports[`LocizeClient should get namespace 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.locize.io/projectid/latest/en/ns1",
      Object {
        "headers": Headers {
          "map": Object {
            "accept": "application/json",
            "authorization": "Bearer apikey",
            "content-type": "application/json",
          },
        },
        "method": "GET",
      },
    ],
  ],
}
`;

exports[`LocizeClient should replace 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.locize.io/update/projectid/latest/en/ns1?replace=true",
      Object {
        "body": "{\\"res1\\":\\"val\\"}",
        "headers": Headers {
          "map": Object {
            "accept": "application/json",
            "authorization": "Bearer apikey",
            "content-type": "application/json",
          },
        },
        "method": "POST",
      },
    ],
  ],
}
`;

exports[`LocizeClient should update modified 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://api.locize.io/update/projectid/latest/en/ns1",
      Object {
        "body": "{\\"res1\\":\\"val\\"}",
        "headers": Headers {
          "map": Object {
            "accept": "application/json",
            "authorization": "Bearer apikey",
            "content-type": "application/json",
          },
        },
        "method": "POST",
      },
    ],
  ],
}
`;
